metaflow/plugins/resources_decorator.py

```diff
@@ -0,0 +1,69 @@
+    This decorator passes this information along when requesting resources
+    to execute this step.
+
+    This decorator can be used for two purpose:
+    When using with AWS Batch decorator, only 'cpu', 'gpu', and 'memory' parameters are supported.
+    When using for Kubeflow Pipeline, 'cpu', and 'memory' sets resource requests;
+        'cpu_limit', 'gpu', 'memory_limit' sets resource limit.
+        For more details please refer to
+        https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+    This decorator is ignored if the execution of the step does not happen on Batch.
+    To use, annotate your step as follows:
+    ```
+    @resources(cpu=32, memory="15G")
+    @step
+    def myStep(self):
+        ...
+    ```
+
+    Parameters
+    ----------
+    cpu : Union[int, float, str]
+        AWS Batch: Number of CPUs required for this step. Defaults to 1. Must be integer.
+        KFP: Number of CPUs required for this step. Defaults to 1. Can input int, float, or str.
+            Support millicpu requests using float or string ending in 'm'.
+            A request with a decimal point, like 0.1, is converted to 100m by the API,
+            and precision finer than 1m is not allowed.
+    cpu_limit : Union[int, float, str]
+        Not for AWS Batch.
+        KFP: Number of CPUs limited for this step. Defaults None - relying on Kubernetes defaults.
+    gpu : int
+        AWS Batch: Number of GPUs required for this step. Defaults to 0.
+        KFP: GPU limit for this step. Defaults to 0.
+            GPU are only supposed to be specified in limit section.
+            See https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/
+    gpu_vendor : str
+        Not for AWS Batch.
+        KFP: "nvidia" or "amd". Defaults to "nvidia".
+    memory : Union[int, str]
+        AWS Batch: Memory size (in MB) required for this step. Defaults to 4000.
+        KFP: Memory size required for this step. Default to 4000 MB. For int input, unit defaults to MB.
+            More memory units are supported, including "E", "P", "T", "G", "M", "K". (i.e. "4000M")
+    memory_limit : Union[int, str]
+        Not for AWS Batch.
+        KFP: Memory limit for this step. Default unit is MB.
+            More memory units are supported, including "E", "P", "T", "G", "M", "K". (i.e. "4000M")
+             Defaults None - relying on Kubernetes defaults.
+    """
+    name = 'resources'
+
+    # Actual defaults are set in .aws.batch.batch_decorator.BatchDecorator and
+    # .kfp.kfp.KubeflowPipelines._get_resource_requirements respectively.
+    # The defaults here simply lists accepted attributes.
+    defaults = {
+        # AWS Batch and KFP supported attributes
+        "cpu": None,
+        "gpu": None,
+        "memory": None,
+
+        # Only KFP supported attributes
+        "cpu_limit": None,
+        "gpu_vendor": None,
+        "memory_limit": None,
```

Comment:
Extending the thinking in the [comment](https://github.com/zillow/metaflow/pull/20/files#r497252347), these three fields should be moved to kfp_decorator as well.

Reply:
Are you suggesting defining another resource decorator? These fields are added here so that `InvalidDecoratorAttribute` exception is not raised from here https://github.com/Netflix/metaflow/blob/master/metaflow/decorators.py#L90-L96

Reply:
We'd likely rename our https://github.com/zillow/metaflow/blob/1b861af0b9c6f81522b6e53ebf6ca88860ef5f1f/metaflow/plugins/kfp/kfp_decorator.py#L16 decorator from `kfp_internal` to `kfp`.

```python
@kfp(cpu_limit = 10)
@step
def hello(self)
   pass
```

Awesome, the confusion might be that one cannot just add `python hello.py --with:kfp` to have the task run on KFP, using local scheduler (runtime.py) for orchestration.  We'd error out saying one must use the KFP scheduler to run a task on KFP:
```
python hello.py kfp run
```
which by default would add `--with:kfp` https://github.com/zillow/metaflow/blob/1b861af0b9c6f81522b6e53ebf6ca88860ef5f1f/metaflow/plugins/kfp/kfp.py#L296

Reply:
@talebzeghmi I wasn't sure what the kfp_internal decorator was for. If it's the `kfp` decorator I'd just add resource handling implementation there, similar to how `@batch` decorator was used. Using `kfp` decorator with make kfp plugin experience much more inline with other Metaflow usages.

Reply:
@talebzeghmi supporting usage of `@kfp` means we allow some steps of the flow running on KFP, and possibly other steps elsewhere. This is not supported by current code base, and is non-trivial considering how our implementation goes right now. 

I will keep the current implementation for resource requirements until we are ready for `@kfp` decorator.

Reply:
@savingoyal Shall we make this resource decorator more generic and support multiple platforms? I personally feel adding `*_limit` makes sense, but may result in additional attribute whitelisting in platform decorators (`@batch`, `@kfp` and in future `@k8s`).

Reply:
Let me know if you wish otherwise. I'm to keep current implementation for now, but we can change at any time.

Reply:
Sure, we can punt this change for later.